Add Azure Key Vault Article (#3861)
umbraco/UmbracoDocs: Extending/Key-Vault/index.md

---
versionFrom: 9.0.0
state: complete
meta.Title: "Azure Key Vault"
meta.Description: "A guide configuring an Azure Key Vault"
---

# Configuring an Azure Key Vault

> From security perspective it's always a good solution to store your application secrets such as a connectionSting and others keys in a Azure Key Vault. This article tells you how to configure your application so that it's ready to use a Key Vault

Reviewer: ```suggestion
From a security perspective, it's always a good solution to store your application secrets such as a connection string and others keys in an Azure Key Vault. This article tells you how to configure your application so that it's ready to use a Key Vault
```

Author: @jonat123 Thanks for the suggestion, just updated the PR.

## Installing the package

Before you begin, you need to install the `Microsoft.Extensions.Configuration.AzureKeyVault` NuGet package. There are two approaches to installing the package:

1. Use your favorite IDE and open up the NuGet Package Manager to search and install the package
1. Use the command line to install the package

## Installing through command line

Navigate to your project folder, which is the folder that contains your .csproj file. Now use the following dotnet add package command to install the package:

```
dotnet add package Microsoft.Extensions.Configuration.AzureKeyVault
```

## Configuration
The next step is to add the Azure Key Vault endpoint to the 'appsettings.json' file. 

```json
{
  "AzureKeyVaultEndpoint": "https://{your-key-vault-name}.vault.azure.net",
}
```

> After adding the Key Vault endpoint you have to update the`CreateHostBuilder` method which you can find in the `Program.cs` class. 

Reviewer: ```suggestion
After adding the Key Vault endpoint you have to update the `CreateHostBuilder` method which you can find in the `Program.cs` class. 
```

Author: @jonat123 Thanks for your suggestion, just updated 👍

```csharp
public static IHostBuilder CreateHostBuilder(string[] args) =>
    Host.CreateDefaultBuilder(args)
        .ConfigureLogging(x => x.ClearProviders())
        .ConfigureAppConfiguration((context, config) =>
        {
            var settings = config.Build();
            var keyVaultEndpoint = settings["AzureKeyVaultEndpoint"];
            if (!string.IsNullOrWhiteSpace(keyVaultEndpoint))
            {
                var azureServiceTokenProvider = new AzureServiceTokenProvider();
                var keyVaultClient =
                    new KeyVaultClient(
                        new KeyVaultClient.AuthenticationCallback(azureServiceTokenProvider
                            .KeyVaultTokenCallback));
                config.AddAzureKeyVault(keyVaultEndpoint, keyVaultClient, new DefaultKeyVaultSecretManager());
            }
        })
        .ConfigureWebHostDefaults(webBuilder => webBuilder.UseStartup<Startup>());
```

## Authentication 

There are several ways to access the Azure Key Vault.

> ### Local Developement 

Reviewer: I feel like this section could do with a bit more information on how you can access the key vault with visual studio and how you store your account into the credential cache 🤔

1. Sign in to Visual Studio using the credentials that can access the Key Vault. 

> 1. Use az CLI to store your preferred account into the credential cache.

Reviewer: ```suggestion
1. Use a CLI to store your preferred account into the credential cache.
```

Author: @jonat123 This one was intended, I meant the Azure CLI (az command), I changed it to Azure CLI to be a bit clearer. 😄

---
versionFrom: 9.0.0
state: complete
meta.Title: "Azure Key Vault"
meta.Description: "A guide configuring an Azure Key Vault"
---

# Configuring an Azure Key Vault

> From a security perspective, it's always a good solution to store your application secrets such as a connection string and others keys in an Azure Key Vault. This article tells you how to configure your application so that it's ready to use a Key Vault

Reviewer 2: ```suggestion
From a security perspective, it's always a good solution to store your application secrets (such as a connection string and others keys) in an Azure Key Vault. This article tells you how to configure your application so that it's ready to use a Key Vault
```

## Installing the package

Before you begin, you need to install the `Microsoft.Extensions.Configuration.AzureKeyVault` NuGet package. There are two approaches to installing the package:

1. Use your favorite IDE and open up the NuGet Package Manager to search and install the package
1. Use the command line to install the package

## Installing through command line

> Navigate to your project folder, which is the folder that contains your .csproj file. Now use the following dotnet add package command to install the package:

Reviewer 2: ```suggestion
Navigate to your project folder, which is the folder that contains your .csproj file. Now use the following 'dotnet add package' command to install the package:
```
